src/Runtime/Core/Maths/Series.cs

```diff
@@ -168,16 +268,38 @@ public Series(uint precisionBits)
                 TetragammaAt1 = new TruncatedPowerSeries(tetragammaAt1Coefficients.ToArray());
                 TetragammaAsymptotic = new TruncatedPowerSeries(tetragammaAsymptoticCoefficients.ToArray());
                 GammalnAsymptotic = new TruncatedPowerSeries(gammalnAsymptoticCoefficients.ToArray());
+                LogExpMinus1RatioAt0 = new TruncatedPowerSeries(logExpMinus1RatioAt0Coefficients.ToArray());
+                NormcdflnAsymptotic = new TruncatedPowerSeries(normcdflnAsymptoticCoefficients.ToArray());
             }
             DigammaAt2 = TruncatedPowerSeries.Generate(GammaAt2.Coefficients.Length, i => GammaAt2.Coefficients[i] * (i + 1));
             Log1Plus = new PowerSeries(Log1PlusCoefficient);
             Log1Minus = new PowerSeries(Log1MinusCoefficient);
+            ExpMinus1 = new PowerSeries<double>(ExpMinus1Coefficient);
+            ExpMinus1RatioMinus1RatioMinusHalf = new PowerSeries<double>(ExpMinus1RatioMinus1RatioMinusHalfCoefficient);
         }
 
         #region Coefficient generators
 
         private static double Log1PlusCoefficient(int n) => n == 0 ? 0.0 : (n % 2 == 0 ? -1.0 : 1.0) / n;
         private static double Log1MinusCoefficient(int n) => n == 0 ? 0.0 : -1.0 / n;
+        private static CoefficientGenerationResult<double> ExpMinus1Coefficient(int n, double prev)
```

Comment:
A C# Enumerator would be simpler.

Reply:
It would be, indeed. I tried it. But, as I explain a doc comment for `CoefficientGenerator` (lines 166-167), "for small coefficient generation functions, like `Series.ExpMinus1Coefficient`, manually calling this delegate is about 2.5 times faster that using `IEnumerable`."

2.5 times performance loss in series evaluation didn't seem acceptable to me, so I sacrificed some simplicity here.

Reply:
I suspect that will vary with platform.  Please add a test that makes this comparison and asserts this claim.

Reply:
Interesting. It indeed varies with platform. More than that - now I can't reproduce on my laptop results I obtained on a machine in the lab. On my laptop performance is opposite of what I said earlier on net461, i.e. IEnumerable is faster. On netcore both 2.1 and 3.1 delegates are ~20% faster than enumerable.

So on Monday I'll recheck what happens on the machine in the lab, and if the results will be consistent with what I've got today, I think we should switch to enumerables, at least until function pointers will make their way into C#.

Reply:
It turned out, my initial benchmark on the lab machine contained an error, so disregard that.

After some research, I found out that the code currently checked in evaluates series 40-60% longer than an implementation that uses enumerables on 32 bit net461, and 10-20% faster on 64 bit net461 and netcore. What surprised me, however, was that the delegate-based stateless coefficient computation used for log(1 +/- x) was consistently 40% faster that enumerable-based. It turned out, that passing the coefficient computation state by value is very expensive and not optimized away on 32-bit net461. When I tried doing so by reference, I managed to get performance consistently better than that of enumerables by 15-30%. Here's all timings:

<table>
<tr>
<th rowspan="2">platform</th><th rowspan="2">approach</th><th colspan="4">function (number of evaluations)</th>
</tr>
<tr>
<th>exp(x) - 1 (5M)</th><th>((exp(x) - 1)/x - 1) / x - 0.5 (5M)</th><th>log(1 + x) (2M)</th><th>log(1 - x) (2M)</th>
</tr>
<tr>
<td rowspan="4">net461, 32 bit</td><td>delegate-based, state passed by value</td>
<td>1687.1197ms</td><td>1588.7696ms</td><td>n/a</td><td>n/a</td>
</tr>
<tr>
<td>delegate-based, state passed by reference</td>
<td>883.5262ms</td><td>819.828ms</td><td>n/a</td><td>n/a</td>
</tr>
<tr>
<td>delegate-based, stateless</td>
<td>n/a</td><td>n/a</td><td>744.9177ms</td><td>666.0605ms</td>
</tr>
<tr>
<td>enumerable-based</td>
<td>924.5574ms</td><td>1137.8162ms</td><td>1842.075ms</td><td>1146.3931ms</td>
</tr>
<tr>
<td rowspan="4">net461, 64 bit</td><td>delegate-based, state passed by value</td>
<td>789.425ms</td><td>676.7761ms</td><td>n/a</td><td>n/a</td>
</tr>
<tr>
<td>delegate-based, state passed by reference</td>
<td>692.4225ms</td><td>675.4358ms</td><td>n/a</td><td>n/a</td>
</tr>
<tr>
<td>delegate-based, stateless</td>
<td>n/a</td><td>n/a</td><td>671.9779ms</td><td>600.6058ms</td>
</tr>
<tr>
<td>enumerable-based</td>
<td>808.1142ms</td><td>932.2971ms</td><td>1044.875ms</td><td>1004.8244ms</td>
</tr>
<tr>
<td rowspan="4">netcore 3.1</td><td>delegate-based, state passed by value</td>
<td>693.2037ms</td><td>570.8964ms</td><td>n/a</td><td>n/a</td>
</tr>
<tr>
<td>delegate-based, state passed by reference</td>
<td>666.51ms</td><td>639.5023ms</td><td>n/a</td><td>n/a</td>
</tr>
<tr>
<td>delegate-based, stateless</td>
<td>n/a</td><td>n/a</td><td>674.3002ms</td><td>601.903ms</td>
</tr>
<tr>
<td>enumerable-based</td>
<td>853.9221ms</td><td>954.0337ms</td><td>1088.9119ms</td><td>1044.6957ms</td>
</tr>
</table>

So, @tminka do you think these 15-30% of performance would justify picking delegates over enumerables? If so, I'll push my new implementation with delegates and the test you mentioned, otherwise, I'll push the implementation with enumerables.

Reply:
I'd like to see what the two implementations look like and run the comparisons myself, which is why I asked you to make a test for it.

Reply:
There's another option to specialize the entire loop around a particular Taylor series, as it was done in various places in the code.  How does that compare?

Reply:
OK, I pushed the new implementation and test, that contains enumerable-based implementation.

About specializing the entire loop, I'm not sure I understand, what do you mean. Something like
```csharp
do
{
    previousApproximation = currentApproximation;
    currentApproximation = GetNextApproximation(); // Specialized part
}
while (previousApproximation != currentApproximation)
```
?

Reply:
[Example of a specialized loop](https://github.com/dotnet/infer/blob/52ee96f4151d1d00a9e4600dcdf3e8da27d86a51/src/Runtime/Core/Maths/SpecialFunctions.cs#L86)

Reply:
After our last conversation it became apparent that all series implementations present before the latest update to this PR were too slow when compared to explicitly evaluating polynomials in code. So, here's another attempt at generalization - building and compiling specialized LINQ expressions. The loss of performance they entail is one additional virtual call, that, currently, can not be optimized away by JIT (unlike non-virtual calls), which is visible on very short polynomials.
In exchange we get some generality when working with power series and better performance for longer series coefficients for which used to be stored in arrays (see CompiledExpressionOnGammaLnPerformanceTest). There's also a potential performance gain in constructing optimized expression trees for approximating infinite series.

@tminka does this sound like a good idea to you?

Reply:
I'm seeing a 30% overhead of the Expression approach, presumably due to the virtual call.  Also, it makes difference where the function is located.  Functions in the test assembly run slower:
Explicit time:  679.3632ms.
Compiled expression-based time: 156.9404ms.
Explicit in MMath time: 118.7893ms.

Reply:
Not for me. On my machine they run faster:
net461
Explicit time:  143.4585ms.
Compiled expression-based time: 214.044ms.
Explicit in MMath time: 170.8941ms.
netcoreapp2.1:
Explicit time:  148.4252ms.
Compiled expression-based time: 282.1745ms.
Explicit in MMath time: 178.0747ms.

Reply:
It is going to be hard to reach consistent conclusions when the location of a function makes a difference.

Reply:
When I inspect the generated assembly code in Visual Studio, it is different depending on whether the function is in Runtime or the Tests project or the TestApp project.  This explains most of the speed difference.  I'm seeing a smaller speed difference between functions with the same assembly code.  This is likely due to cache effects.

Reply:
Interesting. I found another thing that really bothered me: when I run
```csharp
double explicitTimeInms2 = MeasureRunTime(() => { for (int j = 0; j < evaluations; ++j) MMath.ExpMinus1Explicit(point); }, samples);
output.WriteLine($"Explicit in MMath time:\t{explicitTimeInms2}ms.");
explicitTimeInms2 = MeasureRunTime(() => { for (int j = 0; j < evaluations; ++j) MMath.ExpMinus1Explicit(point); }, samples);
output.WriteLine($"Explicit in MMath time:\t{explicitTimeInms2}ms.");
```
on my machine (lambdas are same) I get significantly different results. And when I run different tests, results sometimes depend on the order of the tests. So there's something wrong with the methodology of the benchmarks I hacked.

So I tried benchmarking with [BenchmarkDotNet](https://github.com/dotnet/BenchmarkDotNet), which appears to be a standard tool for benchmarking on .net, and it doesn't seem to have either problem on my machine. I'll push these changes soon and then try running on a VM.

Reply:
I figured out why the assembly code was different: Tests.csproj never set Optimize=true, even in the Release configuration.  That also means that the timing loops were not being optimized.

Reply:
Switched to explicit, inlined, auto-generated series, as discussed.